File: backend/app/api/v1/consent.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.api.schemas.consent import ConsentCreate, ConsentResponse, OnboardingCompleteRequest, ConsentRevokeRequest
from app.domain.repositories.consent_repository import ConsentRepository
from app.api.auth_mode import get_request_user_id
from app.api.router_factory import make_v1_router

router = make_v1_router(prefix="/api/v1/consent", tags=["consent"])


@router.get("", response_model=ConsentResponse)
def get_consent(
    user_id: int = Depends(get_request_user_id),
    db: Session = Depends(get_db),
):
    """Get latest consent record for user"""
    repo = ConsentRepository(db)
    consent = repo.get_latest(user_id)
    if not consent:
        raise HTTPException(status_code=404, detail="No consent record found")
    return consent


@router.post("", response_model=ConsentResponse)
def create_consent(
    payload: ConsentCreate,
    user_id: int = Depends(get_request_user_id),
    db: Session = Depends(get_db),
):
    """
    Create consent record.
    
    All checkboxes must be true for consent to be valid.
    """
    if not all([
        payload.understands_not_medical_advice,
        payload.consents_to_data_analysis,
        payload.understands_recommendations_experimental,
        payload.understands_can_stop_anytime,
    ]):
        raise HTTPException(
            status_code=400,
            detail="All consent checkboxes must be checked"
        )

    repo = ConsentRepository(db)
    consent = repo.create(
        user_id=user_id,
        consent_version=payload.consent_version,
        understands_not_medical_advice=payload.understands_not_medical_advice,
        consents_to_data_analysis=payload.consents_to_data_analysis,
        understands_recommendations_experimental=payload.understands_recommendations_experimental,
        understands_can_stop_anytime=payload.understands_can_stop_anytime,
        consents_to_whoop_ingestion=payload.consents_to_whoop_ingestion,
        consents_to_fitbit_ingestion=payload.consents_to_fitbit_ingestion,
        consents_to_oura_ingestion=payload.consents_to_oura_ingestion,
        consent_text_json=payload.consent_text_json,
    )
    return consent


@router.post("/revoke", response_model=ConsentResponse)
def revoke_consent(
    payload: ConsentRevokeRequest,
    user_id: int = Depends(get_request_user_id),
    db: Session = Depends(get_db),
):
    """
    Revoke consent for the authenticated user.
    
    WEEK 2: Immediately blocks all future provider ingestion.
    """
    repo = ConsentRepository(db)
    consent = repo.revoke(user_id=user_id, reason=payload.reason)
    if not consent:
        raise HTTPException(status_code=404, detail="No consent record found to revoke")
    return consent


@router.post("/complete-onboarding", response_model=ConsentResponse)
def complete_onboarding(
    payload: OnboardingCompleteRequest,
    user_id: int = Depends(get_request_user_id),
    db: Session = Depends(get_db),
):
    """Mark onboarding as completed"""
    repo = ConsentRepository(db)
    consent = repo.mark_onboarding_completed(user_id)
    if not consent:
        raise HTTPException(status_code=404, detail="No consent record found. Please create consent first.")
    return consent

File: frontend/src/api/consent.ts
/**
 * ALPHA WIRING: Use unified API client
 */
import apiClient from "./client";

export interface ConsentData {
  understands_not_medical_advice: boolean;
  consents_to_data_analysis: boolean;
  understands_recommendations_experimental: boolean;
  understands_can_stop_anytime: boolean;
}

export interface ConsentResponse {
  id: number;
  user_id: number;
  consent_version: string;
  consent_timestamp: string;
  understands_not_medical_advice: boolean;
  consents_to_data_analysis: boolean;
  understands_recommendations_experimental: boolean;
  understands_can_stop_anytime: boolean;
  onboarding_completed: boolean;
  onboarding_completed_at: string | null;
}

export async function getConsent(userId: number): Promise<ConsentResponse | null> {
  try {
    const res = await apiClient.get("/consent", {
      params: { user_id: userId },
    });
    return res.data;
  } catch (error: any) {
    if (error.response?.status === 404) {
      return null;
    }
    throw error;
  }
}

export async function createConsent(userId: number, consent: ConsentData): Promise<ConsentResponse> {
  const res = await apiClient.post(
    "/consent",
    {
      consent_version: "1.0",
      ...consent,
    },
    {
      params: { user_id: userId },
    }
  );
  return res.data;
}

export async function completeOnboarding(userId: number): Promise<ConsentResponse> {
  const res = await apiClient.post(
    "/consent/complete-onboarding",
    {},
    {
      params: { user_id: userId },
    }
  );
  return res.data;
}

